fix(authz): optimize subject mapping evaluation performance

strantalis opened the pull request:
### Proposed Changes

Add performance optimizations to subject mapping evaluation hot paths:

  Flattening: O(1) selector lookups via index map

  - GetFromFlattened_Large: 43x faster (228ns → 5.3ns), zero allocs
  - GetFromFlattened_MultipleLookups: 18x faster (509ns → 28ns), zero allocs
  - Index only built for structures ≥8 items; smaller structures use fast linear scan

  EvaluateCondition: Optimized IN/NOT_IN evaluation

  - EvaluateCondition_IN_Large: 57% faster (2.1µs → 0.9µs)
  - EvaluateCondition_IN_Hit: 79% faster (433ns → 93ns), zero allocs
  - Small value sets (≤4) use linear search to avoid map allocation

  Per-entity caches for SubjectSet/ConditionGroup evaluation

  - RepeatedSubjectSets: 48% faster, 45% fewer allocs
  - Cached variants used only in batch evaluation; single-call exports avoid cache overhead

  Action deduplication using per-valueFQN accumulator

  - WithActions_LargeAttrCount: 22% faster, 57% fewer allocations

  Overall impact

  - -38% average improvement in subject mapping evaluation time
  - -66% average improvement in flattening operations
  - Significant allocation reductions across all complex benchmarks

  Trade-offs

  - Flatten_NestedEntity: +28% slower for large structures (index building cost)
  - This cost is amortized over multiple GetFromFlattened calls per entity (break-even: ~3 lookups)
  - Small structures have zero overhead due to index threshold


```Subject Mapping Benchstat
goos: darwin
goarch: arm64
pkg: github.com/opentdf/platform/service/internal/subjectmappingbuiltin
cpu: Apple M4 Max
                                                     │  sm-old.txt  │             sm-new.txt              │
                                                     │    sec/op    │   sec/op     vs base                │
EvaluateCondition_IN_Large-16                          2132.5n ± 4%   917.5n ± 0%  -56.98% (p=0.000 n=20)
EvaluateCondition_IN_Hit-16                            433.25n ± 2%   92.76n ± 2%  -78.59% (p=0.000 n=20)
EvaluateSubjectMappings_LargeAttrCount-16               14.54µ ± 1%   11.27µ ± 1%  -22.49% (p=0.000 n=20)
EvaluateSubjectMappings_RepeatedSubjectSets-16         11.049µ ± 0%   5.784µ ± 1%  -47.66% (p=0.000 n=20)
EvaluateSubjectMappingsWithActions_LargeAttrCount-16    38.43µ ± 1%   30.06µ ± 1%  -21.78% (p=0.000 n=20)
EvaluateSubjectMappingMultipleEntitiesWithActions-16    162.4µ ± 2%   143.2µ ± 1%  -11.83% (p=0.000 n=20)
EvaluateSubjectSet-16                                   65.28n ± 1%   55.27n ± 1%  -15.33% (p=0.000 n=20)
EvaluateSubjectMappings-16                              1.111µ ± 1%   1.061µ ± 1%   -4.50% (p=0.000 n=20)
geomean                                                 4.013µ        2.473µ       -38.36%

                                                     │   sm-old.txt   │                sm-new.txt                 │
                                                     │      B/op      │     B/op      vs base                     │
EvaluateCondition_IN_Large-16                            960.0 ± 0%      1832.0 ± 0%   +90.83% (p=0.000 n=20)
EvaluateCondition_IN_Hit-16                              960.0 ± 0%         0.0 ± 0%  -100.00% (p=0.000 n=20)
EvaluateSubjectMappings_LargeAttrCount-16              35.70Ki ± 0%     18.85Ki ± 0%   -47.19% (p=0.000 n=20)
EvaluateSubjectMappings_RepeatedSubjectSets-16         28.42Ki ± 0%     12.36Ki ± 0%   -56.49% (p=0.000 n=20)
EvaluateSubjectMappingsWithActions_LargeAttrCount-16   62.20Ki ± 0%     38.55Ki ± 0%   -38.03% (p=0.000 n=20)
EvaluateSubjectMappingMultipleEntitiesWithActions-16   253.1Ki ± 0%     217.1Ki ± 0%   -14.22% (p=0.000 n=20)
EvaluateSubjectSet-16                                    0.000 ± 0%       0.000 ± 0%         ~ (p=1.000 n=20) ¹
EvaluateSubjectMappings-16                             1.634Ki ± 0%     1.634Ki ± 0%         ~ (p=1.000 n=20) ¹
geomean                                                             ²                 ?                       ² ³
¹ all samples are equal
² summaries must be >0 to compute geomean
³ ratios must be >0 to compute geomean

                                                     │  sm-old.txt   │                sm-new.txt                │
                                                     │   allocs/op   │  allocs/op   vs base                     │
EvaluateCondition_IN_Large-16                           4.000 ± 0%      3.000 ± 0%   -25.00% (p=0.000 n=20)
EvaluateCondition_IN_Hit-16                             4.000 ± 0%      0.000 ± 0%  -100.00% (p=0.000 n=20)
EvaluateSubjectMappings_LargeAttrCount-16               266.0 ± 0%      145.0 ± 0%   -45.49% (p=0.000 n=20)
EvaluateSubjectMappings_RepeatedSubjectSets-16          233.0 ± 0%      129.0 ± 0%   -44.64% (p=0.000 n=20)
EvaluateSubjectMappingsWithActions_LargeAttrCount-16    709.0 ± 0%      302.0 ± 0%   -57.40% (p=0.000 n=20)
EvaluateSubjectMappingMultipleEntitiesWithActions-16   3.269k ± 0%     2.010k ± 0%   -38.51% (p=0.000 n=20)
EvaluateSubjectSet-16                                   0.000 ± 0%      0.000 ± 0%         ~ (p=1.000 n=20) ¹
EvaluateSubjectMappings-16                              30.00 ± 0%      30.00 ± 0%         ~ (p=1.000 n=20) ¹
geomean                                                            ²                ?                       ² ³
¹ all samples are equal
² summaries must be >0 to compute geomean
³ ratios must be >0 to compute geomean
```

```Flatten Benchstat
goos: darwin
goarch: arm64
pkg: github.com/opentdf/platform/lib/flattening
cpu: Apple M4 Max
                                    │ flatten-old.txt │           flatten-new.txt           │
                                    │     sec/op      │   sec/op     vs base                │
GetFromFlattened_Large-16               228.500n ± 2%   5.331n ± 1%  -97.67% (p=0.000 n=20)
GetFromFlattened_MultipleLookups-16      508.80n ± 2%   27.90n ± 5%  -94.52% (p=0.000 n=20)
Flatten_NestedEntity-16                   2.269µ ± 2%   2.917µ ± 1%  +28.54% (p=0.000 n=20)
FlattenMap-16                             224.0n ± 1%   221.0n ± 3%        ~ (p=0.351 n=20)
FlattenMapWithinMap-16                    374.0n ± 1%   374.9n ± 1%        ~ (p=0.372 n=20)
GetFromFlattened-16                       13.68n ± 2%   13.87n ± 1%        ~ (p=0.163 n=20)
geomean                                   259.1n        89.04n       -65.63%

                                    │ flatten-old.txt │              flatten-new.txt              │
                                    │      B/op       │     B/op      vs base                     │
GetFromFlattened_Large-16                  16.00 ± 0%      0.00 ± 0%  -100.00% (p=0.000 n=20)
GetFromFlattened_MultipleLookups-16        80.00 ± 0%      0.00 ± 0%  -100.00% (p=0.000 n=20)
Flatten_NestedEntity-16                  5.861Ki ± 0%   7.714Ki ± 0%   +31.61% (p=0.000 n=20)
FlattenMap-16                              360.0 ± 0%     360.0 ± 0%         ~ (p=1.000 n=20) ¹
FlattenMapWithinMap-16                     624.0 ± 0%     624.0 ± 0%         ~ (p=1.000 n=20) ¹
GetFromFlattened-16                        16.00 ± 0%     16.00 ± 0%         ~ (p=1.000 n=20) ¹
geomean                                    173.9                      ?                       ² ³
¹ all samples are equal
² summaries must be >0 to compute geomean
³ ratios must be >0 to compute geomean

                                    │ flatten-old.txt │             flatten-new.txt              │
                                    │    allocs/op    │  allocs/op   vs base                     │
GetFromFlattened_Large-16                  1.000 ± 0%    0.000 ± 0%  -100.00% (p=0.000 n=20)
GetFromFlattened_MultipleLookups-16        5.000 ± 0%    0.000 ± 0%  -100.00% (p=0.000 n=20)
Flatten_NestedEntity-16                    94.00 ± 0%   117.00 ± 0%   +24.47% (p=0.000 n=20)
FlattenMap-16                              11.00 ± 0%    11.00 ± 0%         ~ (p=1.000 n=20) ¹
FlattenMapWithinMap-16                     17.00 ± 0%    17.00 ± 0%         ~ (p=1.000 n=20) ¹
GetFromFlattened-16                        1.000 ± 0%    1.000 ± 0%         ~ (p=1.000 n=20) ¹
geomean                                    6.668                     ?                       ² ³
¹ all samples are equal
² summaries must be >0 to compute geomean
³ ratios must be >0 to compute geomean
```
[sm-old.txt](https://github.com/user-attachments/files/23963540/sm-old.txt)
[flatten-old.txt](https://github.com/user-attachments/files/23963541/flatten-old.txt)
[flatten-new.txt](https://github.com/user-attachments/files/23963542/flatten-new.txt)
[sm-new.txt](https://github.com/user-attachments/files/23963543/sm-new.txt)

### Checklist

- [ ] I have added or updated unit tests
- [ ] I have added or updated integration tests (if appropriate)
- [ ] I have added or updated documentation

### Testing Instructions

strantalis committed:
perf(authz): optimize subject mapping evaluation performance

Add performance optimizations to subject mapping evaluation hot paths:

- Flattening: Add internal index map for O(1) selector lookups
  - GetFromFlattened_Large: 39x faster (217ns → 5.5ns), zero allocs
  - GetFromFlattened_MultipleLookups: 17x faster, zero allocs

- EvaluateCondition: Set-based IN/NOT_IN with early returns
  - EvaluateCondition_IN_Large: 2.3x faster
  - EvaluateCondition_IN_Hit: 4.6x faster, zero allocs

- Per-entity caches for SubjectSet/ConditionGroup evaluation
  - RepeatedSubjectSets: 21% faster, 37% fewer allocs

- Action deduplication using per-valueFQN accumulator
  - WithActions: 6% faster, 51% fewer allocations

- HierarchyRule: O(h) definition scan vs O(e) entitlements scan

Trade-off: Flatten() is ~2x slower due to index building, but this
cost is amortized over many GetFromFlattened calls per entity.

strantalis committed:
perf(authz): fix selector dedupe in JIT PDP

Fixed bug where subjectPropertySet was checked but never updated,
causing duplicate selectors to be added.

strantalis committed:
feat(examples): add complex authorization benchmarks

Add new benchmark commands that exercise subject mapping evaluation
with multiple attributes per resource and multiple resources per request:

- benchmark-decision-complex: v2 API with configurable resources/attrs
- benchmark-decision-v1-complex: v1 API with same capabilities

These benchmarks show where optimization improvements shine by testing
with 500 resources × 5 attributes = 2,500 attribute checks per request.

Also updates CI workflow to run and report these benchmarks.

strantalis committed:
revert(examples): remove complex authorization benchmarks

The complex benchmarks were not providing useful metrics as all
decisions were being denied. Removing them to simplify the CI.

strantalis committed:
perf(authz): optimize flatten and subject mapping for small inputs

- Flatten: separate index building into a linear pass, skip index for
  structures with fewer than 8 items (linear scan is faster)
- EvaluateSubjectSet/ConditionGroup: split into cached and non-cached
  versions, use non-cached for single evaluations to avoid map overhead
- IN/NOT_IN conditions: use linear search for small value sets (<=4)
  to avoid map allocation

Benchmarks show -38% geomean improvement in subject mapping evaluation
and -66% geomean improvement in flattening lookups.

strantalis commented:
/gemini review

jakedoublev commented on service/internal/subjectmappingbuiltin/subject_mapping_builtin.go:
```diff
@@ -97,19 +101,24 @@ func EvaluateSubjectMappings(attributeMappings map[string]*attributes.GetAttribu
 		}
+
+		// Per-entity caches to avoid re-evaluating the same subject sets/groups across mappings
+		subjectSetCache := make(map[*policy.SubjectSet]bool)
```
/gemini how can we ensure these caches are garbage collected as quickly/easily as possible when we're done with them? While the time optimization seems valuable, this is more memory than before on a very hot path.

strantalis committed:
test(authz): add test for hierarchyRule defensive code path

Add TestHierarchyRule_DefinitionMissingValues to exercise the defensive
code path that handles data inconsistency where resourceValueFQNs
contains values not present in attrDefinition.GetValues().

This guards against panics if the policy service returns incomplete
attribute definitions.
